### PR TITLE
bpo-1635741: port _ctypes extension module to multiphase initialization 

### PR DESCRIPTION
**Pull request by kumaraditya303**
https://bugs.python.org/issue1635741

### CODE REVIEW

**Comment by kumaraditya303** [comment]
cc @vstinner

File: Modules/_ctypes/_ctypes.c
Comment (shihai1991) [inline Modules/_ctypes/_ctypes.c:5844]: I am just curious about those codes in `_ctypes_mod_exec`, do we need judge the attr manually?
```
_unpickle = PyObject_GetAttrString(mod, "_unpickle");
    if (_unpickle == NULL) {
        return -1;
    }
```

**Review by vstinner — CHANGES_REQUESTED** [review]
You should not convert an extension to the multi-phase init API until all static types defined by the extension are defined as heap types.

See _ctypes_add_types(): there are *many* types defined as static types. Just one example: PyCArg_Type.

**Comment by erlend-aasland** [comment]
> You should not convert an extension to the multi-phase init API until all static types defined by the extension are defined as heap types.

True; but sadly, converting static types to heap types has been put on hold by the SC. The good news is that work is in progress to resolve the "deadlock", so we can continue applying PEP 489 and PEP 630 to the standard library 😃 

See [SC communications as of February 2021](https://github.com/python/steering-council/blob/1d85eefdc5861a096c3859e9990dbc8527c5973b/updates/2021-02-steering-council-update.md):

_"The Steering Council discussed the ongoing work on porting types in the standard library to heap-types and the subinterpreter-related changes. It was decided that through Pablo, the Steering Council will ask the core developers driving those changes to create an informational PEP and not to make any more changes in this area after beta 1, as per our general policy."_

**Comment by vstinner** [comment]
> ask the core developers driving those changes to create an informational PEP

Eric is working on a PEP: https://github.com/python/peps/pull/2212

Not sure if it fulfill SC requirements to convert more static types to heap types.

**Comment by kumaraditya303** [comment]
IIUC converting static types to heap types is a different issue https://bugs.python.org/issue40077 which is not I am tackling in this PR, also itertools module also creates many static types but still uses multiphase initialization https://github.com/python/cpython/blob/be578e0c063dad1dbb273f86d5bc77e4e6f14583/Modules/itertoolsmodule.c#L4907 so how is _ctypes different ?

**Comment by erlend-aasland** [comment]
It is subinterpreter related. Also, we first convert global state to module state, _then_ apply multi-phase init, as Victor already pointed out.

Some modules have been converted to multi-phase init prematurely. That does not alter the status quo, or set a precedent.

**Comment by corona10** [comment]
FYI, Please read https://github.com/vstinner/talks/blob/main/2021-PyconUS/subinterpreters.pdf how we've converted modules :)

**Review by corona10 — CHANGES_REQUESTED** [review]
Same as Victor

**Comment by erlend-aasland** [comment]
FYI: https://github.com/python/steering-council/issues/99

**Comment by erlend-aasland** [comment]
FYI, PEP-687 was just [accepted](https://discuss.python.org/t/pep-687-isolating-modules-in-the-standard-library/14824/4?u=erlendaasland).

**Pull request closed without merging**
